src/PdoConnection.php

```
@@ -58,6 +58,16 @@ class PdoConnection
     private $reuseConnection = true;
 
     /**
+     * @var int Number of open transactions
+     */
+    private $transactionCount = 0;
+
+    /**
+     * @var bool Flag that tells whether a rollback is needed
+     */
+    private $rollbackNeeded = false;
```

Why have a flag that does that same thing as `$transactionCount === 0`? Also the naming is misleading. Just because a transaction has been processed doesn't mean it needs to be rolled back.

---

$transactionCount represents how many transactions are currently open.  $rollbackNeeded is only true when a rollback is called on a transaction within another transaction, which is very different from $transactionCount === 0.  I means that if the outer transaction tries to commit, we need to rollback instead.

---

Performing a rollback within a commit is not a good design choice, imo. If you can't commit, or you can't rollback you probably need to throw an exception because that's definitely programmer error.

Also, if they is just a queue system, why not use a queue?